src/testing/cluster.zig

```diff
@@ -300,28 +305,29 @@ pub fn ClusterType(comptime StateMachineType: fn (comptime Storage: type, compti
             }
         }
 
-        pub fn restart_replica(cluster: *Self, replica_index: u8) void {
+        /// Returns whether the replica was crashed.
+        /// Returns an error when the replica was unable to recover (open).
+        pub fn restart_replica(cluster: *Self, replica_index: u8) !void {
             assert(cluster.replica_health[replica_index] == .down);
 
+            // Pass the old replica's Time through to the new replica. It will continue to tick while
+            // the replica is crashed, to ensure the clocks don't desyncronize too far to recover.
+            var time = cluster.replicas[replica_index].time;
+            try cluster.open_replica(replica_index, time);
             cluster.network.process_enable(.{ .replica = replica_index });
             cluster.replica_health[replica_index] = .up;
         }
 
         /// Reset a replica to its initial state, simulating a random crash/panic.
         /// Leave the persistent storage untouched, and leave any currently
         /// inflight messages to/from the replica in the network.
-        ///
-        /// Returns whether the replica was crashed.
-        /// Returns an error when the replica was unable to recover (open).
-        pub fn crash_replica(cluster: *Self, replica_index: u8) !void {
+        pub fn crash_replica(cluster: *Self, replica_index: u8) void {
             assert(cluster.replica_health[replica_index] == .up);
 
             // Reset the storage before the replica so that pending writes can (partially) finish.
             cluster.storages[replica_index].reset();
 
-            const replica = &cluster.replicas[replica_index];
-            const replica_time = replica.time;
-            replica.deinit(cluster.allocator);
+            cluster.replicas[replica_index].deinit(cluster.allocator);
```

Comment:
Let's add `self.* = undefined` to `Replica.deinit`. I _think_ that would add runtime assertions in debug if we actually try to access that?

Reply:
Ah, that won't work, because we do access Replica to tick it's timer :thinking:

Reply:
We could change the `time` that is passed into a `*Time` (similar to how `Storage` and `MessageBus` are not owned by the replica itself).

Reply:
Then `time` would be owned by the `Cluster` (like storage and messagebus).

Reply:
Accessing a field of a de-inited struct feels a bit questionable, but probably is OK for simulator. The alternative would be to fold `replica_healf` into `replicas`:

```zig
union(enum) {
  up: Replica,
  down: Time,
}
```

but that feels like too much churn. But maybe that's ok?